error_group.go

```
@@ -0,0 +1,103 @@
+package sentry
+
+import (
+	"fmt"
+	"reflect"
+	"slices"
+)
+
+const (
+	MechanismTypeGeneric string = "generic"
+	MechanismTypeChained string = "chained"
+
+	MechanismSourceCause string = "cause"
+)
+
+func convertErrorToExceptions(err error) []Exception {
+	var exceptions []Exception
+	visited := make(map[error]bool)
+	convertErrorDFS(err, &exceptions, nil, "", visited)
+
+	// mechanism type is used for debugging purposes, but since we can't really distinguish the origin of who invoked
+	// captureException, we set it to nil if the error is not chained.
+	if len(exceptions) == 1 {
+		exceptions[0].Mechanism = nil
+	}
+
+	slices.Reverse(exceptions)
+
+	return exceptions
+}
+
+func convertErrorDFS(err error, exceptions *[]Exception, parentID *int, source string, visited map[error]bool) {
+	if err == nil {
+		return
+	}
+
+	if visited[err] {
+		return
+	}
+	visited[err] = true
+
+	var isExceptionGroup bool
+
+	switch err.(type) {
+	case interface{ Unwrap() []error }, interface{ Unwrap() error }, interface{ Cause() error }:
```

I believe this is not correct according to the RFC and to how other SDKs work.
I see you've added a workaround on line 91 to handle some specific cases (which I haven't fully though about) but the problem still stands.

The correct thing here would be to set `isExceptionGroup = true` only when we have `Unwrap() []error`.
That really matches the concept of exception **group**. 
In other cases, what you would have is a **chain**.

Please try the following example:
```go
	deepErr := errors.New("root cause")
	for i := 1; i <= 5; i++ {
		deepErr = fmt.Errorf("layer %d: %w", i, deepErr)
	}
	deepErr = fmt.Errorf("top level error: %w", deepErr)
	fmt.Println(deepErr)
	sentry.CaptureException(deepErr)
```

With the current code, you would get this: https://sentry-sdks.sentry.io/issues/6915067203/events/535bf98e4823453fb4ae3cb85fcb9837/

With the "correct" version (change this line to `case interface{ Unwrap []error }:`) you would get this:
https://sentry-sdks.sentry.io/issues/6915069781/?project=4508683222843393&query=&referrer=issue-stream

Note the title difference.
Between the two, I would prefer the second one, because I see the full top-level message.

The good thing about the "incorrect" approach that we have in the current commit I'm reviewing, is that it doesn't break grouping with respect to issues created by older versions of the SDK.
This is because so far the SDK has incorrectly reported chains as groups, see https://github.com/getsentry/sentry-go/blame/8921d3c248d9619dea4bf9f1fe909a76f5982d48/interfaces.go#L477 and the git blame.
In other words, handling chains correctly would create new groups.

So the tradeoff here is between correctness + better titling vs not breaking grouping for existing chains, and I think we should consider it or at least be aware of it.

---

In the end it's what we consider a group vs what we don't in the language, so IMO chained errors can be considered groups. The only thing that changes, behavior wise is that in the UI, the chained errors get hidden (and grouped together), so I was personally in favor of that as an approach. However if it affects titles, we should follow the "correct" way so that the whole title message is included.